src/dawn/Optimizer/PassTemporaryToStencilFunction.cpp

```diff
@@ -0,0 +1,409 @@
+                     temporaryFieldAccessIDToFunctionCall,
+                 const sir::Interval sirInterval,
+                 std::shared_ptr<std::vector<sir::StencilCall*>> stackTrace)
+      : instantiation_(instantiation),
+        temporaryFieldAccessIDToFunctionCall_(temporaryFieldAccessIDToFunctionCall),
+        interval_(sirInterval), stackTrace_(stackTrace) {}
+
+  virtual ~TmpReplacement() {}
+
+  static std::string offsetToString(int a) {
+    return ((a < 0) ? "minus" : "") + std::to_string(std::abs(a));
+  }
+
+  /// @brief create the name of a newly created stencil function associated to a tmp computations
+  std::string makeOnTheFlyFunctionName(const std::shared_ptr<FieldAccessExpr>& expr) {
+    return expr->getName() + "_OnTheFly" + "_i" + offsetToString(expr->getOffset()[0]) + "_j" +
+           offsetToString(expr->getOffset()[1]) + "_k" + offsetToString(expr->getOffset()[2]);
+  }
+
+  unsigned int getNumTmpReplaced() { return numTmpReplaced_; }
+  void resetNumTmpReplaced() { numTmpReplaced_ = 0; }
+
+  /// @brief previsit the access to a temporary. Finalize the stencil function instantiation and
+  /// recompute its <statement,accesses> pairs
+  virtual bool preVisitNode(std::shared_ptr<FieldAccessExpr> const& expr) override {
+    int accessID = instantiation_->getAccessIDFromExpr(expr);
+    if(!temporaryFieldAccessIDToFunctionCall_.count(accessID))
+      return false;
+    // TODO we need to version to tmp function generation, in case tmp is recomputed multiple times
+    std::string callee = expr->getName() + "_OnTheFly";
+    std::shared_ptr<StencilFunctionInstantiation> stencilFun =
+        instantiation_->getStencilFunctionInstantiationCandidate(callee);
+
+    std::string fnClone = makeOnTheFlyFunctionName(expr);
+
+    // the temporary was not replaced by a stencil function by previous visitor, for some reasons it
+    // was not supported
+    if(instantiation_->hasStencilFunctionInstantiation(fnClone))
+      return true;
+
+    DAWN_ASSERT(temporaryFieldAccessIDToFunctionCall_.count(accessID));
+
+    // retrieve the sir stencil function definition
+    std::shared_ptr<sir::StencilFunction> sirStencilFunction =
+        temporaryFieldAccessIDToFunctionCall_.at(accessID).sirStencilFunction_;
+
+    // we create a new sir stencil function, since its name is demangled from the offsets.
+    // for example, for a tmp(i+1) the stencil function is named as tmp_OnTheFly_iplus1
+    std::shared_ptr<sir::StencilFunction> sirStencilFunctionInstance =
+        std::make_shared<sir::StencilFunction>(*sirStencilFunction);
+
+    sirStencilFunctionInstance->Name = fnClone;
+
+    // TODO is this really needed, we only change the name, can we map multiple function
+    // instantiations (i.e. different offsets) to the same sir stencil function
+    // insert the sir::stencilFunction into the StencilInstantiation
+    instantiation_->insertStencilFunctionIntoSIR(sirStencilFunctionInstance);
```

Comment:
The convention for now was that we did not touch the SIR moving trough the passes but rather use it to "initialize" the IIR. I think this is a good idea and we should not try to keep the SIR as well as the IIR up to speed at all times

Reply:
let discuss this. I understand your point, but the problem is that later other passes and code generation will use information of the SIR, not just IIR. So I dont know how to solve the issue your brought